### DISCUSSION TITLE
Overriding test env vars for individual tests

### DISCUSSION THREAD

**Discussion by ndench** [symfony/symfony#43574] [Q&A]
I've come across an issue where I need to test that a service does different things depending on an environment variable (for example, whether it's running in "sandbox" mode for an external integration).

I've found [this SO question](https://stackoverflow.com/q/51692141/1393498) which gives a way of setting and resetting the env var at the start and end of a test, however I was wondering if there is a less error prone way? (If I forget to reset it at the end of the test, then the change will affect other tests being run).

Here are the examples given in the answer:

```php
public function testWithPutenv(): void
{
    // Save what the env is currently
    // You probably want to manually cast it to the type you need
    // This will return false if the env var is not set
    // which can't really be handled very well
    $originalEnv = getenv('MY_ENV_VAR');

    // Set the env to what you need
    putenv('MY_ENV_VAR=foo');

    $client = static::createClient();
    $client->request('POST', '/do-something');

    // assertions...

    // reset the env to what it was before
    putenv(sprintf('MY_ENV_VAR=%s', (string) $originalEnv));
}

public function testWithENV(): void
{
    // Save what the env is currently
    // Have a default if it's not set
    // Alternatively you can check whether it's set or not and 
    // use `unset` to "reset" it at the end of the test
    $originalEnv = $_ENV('MY_ENV_VAR') ?? 'bar';

    // Set the env to what you need
    $_ENV['MY_ENV_VAR'] = 'foo';

    $client = static::createClient();
    $client->request('POST', '/do-something');

    // assertions...

    // reset the env to what it was before
    $_ENV['MY_ENV_VAR'] = $originalEnv;
}

public function testWithENV(): void
{
    // Save what the env is currently
    // Default to empty string, so we know whether we need to unset it
    $originalEnv = $_SERVER('MY_ENV_VAR') ?? '';

    // Set the env to what you need
    $_SERVER['MY_ENV_VAR'] = 'foo';

    $client = static::createClient();
    $client->request('POST', '/do-something');

    // assertions...

    // reset the env to what it was before
    if ('' === $originalEnv) {
        unset($_SERVER['MY_ENV_VAR']);
    } else {
        $_SERVER['MY_ENV_VAR'] = $originalEnv;
    }
}
```

**Comment by michaoj** [comment]
Can't you just define the env var on the command line ?

`MY_ENV_VAR=my_value php ./vendor/bin/phpunit `

This way your env var will only be updated for your test run

**Reply by ndench** [reply]
No unfortunately, because I have two tests that assert different things happen with different environment variables.
eg.

1. Assert that with `SANDBOX=true` my API client hits the https://sandox.app.com/api endpoint
2. Assert that with `SANDBOX=false` my API client hits the https://app.com/api endpoint

I'd rather run all the tests at once with `./vendor/bin/phpunit` instead of having to run each test individulaly.

**Comment by lugosium** [comment]
Have you try https://phpunit.readthedocs.io/en/9.5/annotations.html#appendixes-annotations-runinseparateprocess ?

**Reply by ndench** [reply]
This could work. It should stop me from having to unset the environment variable at the end of the test. Though I still need to remember to put this annotation on the test cases, which is pretty similar to unsetting the environment variable.
